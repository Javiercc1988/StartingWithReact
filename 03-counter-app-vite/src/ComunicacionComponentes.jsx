export const ComunicacionComponents2 = (props) => {
    if(!title){
        throw new Error('El title no existe');
    }
  return (
    <>
      <h1>{props.title}</h1>
      <p>{props.name}</p>
      <p>{props.suma}</p>

    </>
  );
};

// OpciÃ³n desestructurada
export const ComunicacionComponents = ({title, name, suma}) => {

    if(!title){
        throw new Error('El title no existe');
    }

  return (
    <>
      <h1>{title}</h1>
      <p>{name}</p>
      <p>{suma}</p>
    </>
  );
};
