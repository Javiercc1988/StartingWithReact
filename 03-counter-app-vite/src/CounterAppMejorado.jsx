import PropTypes from "prop-types";
import { useState } from "react";

export const CounterAppMejorado = ({ valorInicial = 0 }) => {
  // LÃ³gica
  const [counterValue, setCounter] = useState(valorInicial);

  const incrementarContador = () => setCounter(counterValue + 1);

  const decrementarContador = () => setCounter(counterValue - 1);

  const resetContador = () => setCounter(valorInicial);

  // HTML
  return (
    <>
      <h1>Counter App Mejorado</h1>
      <h3>Valor: {counterValue}</h3>

      <button onClick={incrementarContador}> +1 </button>
      <button onClick={decrementarContador}> -1 </button>
      <button aria-label="btn-reset" onClick={resetContador}>Reset</button>
    </>
  );
};

CounterAppMejorado.propTypes = {
  valorInicial: PropTypes.number.isRequired,
};
