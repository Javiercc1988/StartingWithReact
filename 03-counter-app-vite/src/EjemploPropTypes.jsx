// Es necesario instalarlo con vite ya que no se instala de inicio
// npm i prop-types
// Las proptypes sirven para definir el tipo a las properties del componente
import PropTypes from "prop-types";

export const EjemploPropTypes = ({ title, subtitle, suma}) => {
  return (
    <>
      <h1 data-testid='test-title'>{title}</h1>
      <p>{subtitle}</p>
      <p>{suma}</p>
    </>
  );
};


// De esta forma podemos dar un tipado a las props del componente
EjemploPropTypes.propTypes = {
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string,
  suma: PropTypes.number
};

// Valores por defecto
EjemploPropTypes.defaultProps = {
    // title: 'No hay ningún título',
    subtitle: 'No hay subtitulo',
    suma: 2,

}