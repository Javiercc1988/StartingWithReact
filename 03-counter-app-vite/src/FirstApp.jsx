import { Fragment } from "react";

//1 En react para agrupar dos elementos es necesario tener un nodo padre como podría ser un DIV:
export const FirstAppConDiv = () => {
  return (
    <div>
      <h1> First App React Javi </h1>
      <p>Soy un parrafo</p>
    </div>
  );
};

//2 Existe una opción llamada fragment que debemos importar
export const FirstAppConFragment = () => {
  return (
    <Fragment>
      <h1> First App React Javi </h1>
      <p>Soy un parrafo</p>
    </Fragment>
  );
};

//3 Hay una tercera opción que es la siguiente y más comoda ya que nos ahorra importaciones
// Es importante saber que siempre debe haber un nodo padre cuando queremos concatenar elementos
export const FirstApp = () => {
  return (
    <>
      <h1> First App React Javi </h1>
      <p>Soy un parrafo</p>
    </>
  );
};

const name = "Javi";
const surname = "Pruebas";

const testArray = [1, 2, 3, 4, 5];
const testObject = {
  title: "Titulo 1",
  subtitle: "subtitulo2",
};

const getTitle = () => {
    return testObject.title
}

export const Interpolation = () => {

    // Dejar fuera las constantes de la data y las funciones es interesante ya que de esta forma cuando se deba
    // redibujar el componente no tendrá que procesar estos datos
    // Si no hay una dependencia del componente es mejor dejarlo fuera

  return (
    <>
      <h1> {name} </h1>
      <p>{surname}</p>

      <p>{testArray}</p>

      <p>{getTitle()}</p>
      {/* Imprimir un objeto */}

      <code>{JSON.stringify(testObject)}</code>
    </>
  );
};
