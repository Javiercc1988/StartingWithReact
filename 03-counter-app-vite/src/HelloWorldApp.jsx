export const HelloWorldApp = () => {
  return <h1> Hellow word App </h1>;
};
