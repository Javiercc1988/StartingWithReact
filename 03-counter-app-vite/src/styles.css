*{
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}

html, body {
    font-family: sans-serif;
    color: white;
    background-color: #21232a;
    font-size: 1.3rem;
    padding: 70px;
}

button{
    padding: 15px;
    font-size: 1.3rem;
    border-radius: 10px;
}