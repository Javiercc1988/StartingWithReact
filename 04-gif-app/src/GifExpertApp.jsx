import { useState } from "react";
import { AddCategory } from "./components/AddCategory";
import { GifGrid } from "./components/GifGrid";

export const GifExpertApp = () => {
  const [categories, setCategories] = useState(["Dragon ball"]);

  const onAddCategory = (newCategory) => {
    const existCategory = categories.some(
      (category) => category.toLowerCase() === newCategory.toLowerCase()
    );

    if (existCategory) return;

    setCategories([newCategory, ...categories]);
  };

  return (
    <>
      <h1>GifExpertApp</h1>

      <AddCategory onAddNewCategory={(event) => onAddCategory(event)} />

      {
        categories.map((category) => (
          <GifGrid category={category} />
        ))
      }
    </>
  );
};
