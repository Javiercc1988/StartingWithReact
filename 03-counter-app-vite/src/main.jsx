// IMPORTAR REACT
import React from "react";
// IMPORTAR ALGO PARA RENDERIZARLO (DOM)
import ReactDOM from "react-dom/client";
import { HelloWorldApp } from "./HelloWorldApp";
import { FirstApp, Interpolation } from "./FirstApp";
import { ComunicacionComponents } from "./ComunicacionComponentes";
import { EjemploPropTypes } from "./EjemploPropTypes";
import { CounterAppMejorado } from "./CounterAppMejorado";

// Importar estilos globales
import "./styles.css";
import { CounterApp } from "./CounterApp";

// RENDERIZADO
ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    {/* <HelloWorldApp /> */}
    {/* <FirstApp /> */}
    {/* <Interpolation /> */}
    {/* <ComunicacionComponents
      title="Hola!"
      name="Prueba de props!"
      suma={15 + 15}
    /> */}

    {/* <EjemploPropTypes/> */}

    {/* <CounterApp value={135} /> */}
    <CounterAppMejorado valorInicial={10}/>
  </React.StrictMode>
);
