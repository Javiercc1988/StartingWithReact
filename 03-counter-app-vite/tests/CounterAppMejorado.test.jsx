import { fireEvent, render, screen } from "@testing-library/react";
import { CounterAppMejorado } from "../src/CounterAppMejorado";

describe("Pruebas componente CounterAppMejorado", () => {
  const initialValue = 100;

  test("Debe hacer match con el snapshot", () => {
    const { container } = render(<CounterAppMejorado valorInicial={22} />);
    expect(container).toMatchSnapshot();
  });

  test("El valor inicial debe ser 100", () => {
    render(<CounterAppMejorado valorInicial={initialValue} />);
    expect(screen.getByText(`Valor: ${initialValue}`)).toBeTruthy();
  });

  test("Debe incrementar en 1 el contador", () => {
    render(<CounterAppMejorado valorInicial={initialValue} />);
    fireEvent.click(screen.getByText("+1"));
    expect(screen.getByText("Valor: 101")).toBeTruthy();
  });

  test("debe de decrementar con el botÃ³n -1", () => {
    render(<CounterAppMejorado valorInicial={initialValue} />);
    fireEvent.click(screen.getByText("-1"));
    expect(screen.getByText("Valor: 99")).toBeTruthy();
  });

  test("debe hacer reset", () => {
    render(<CounterAppMejorado valorInicial={initialValue} />);
    fireEvent.click(screen.getByText("+1"));
    fireEvent.click(screen.getByText("+1"));
    fireEvent.click(screen.getByText("+1"));

    fireEvent.click(screen.getByRole("button", { name: "btn-reset" }));
    expect(screen.getByText("Valor: 100")).toBeTruthy();
  });
});
