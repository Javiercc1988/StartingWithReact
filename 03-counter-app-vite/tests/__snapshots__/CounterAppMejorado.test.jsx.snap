// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas componente CounterAppMejorado Debe hacer match con el snapshot 1`] = `
<div>
  <h1>
    Counter App Mejorado
  </h1>
  <h3>
    Valor: 
    22
  </h3>
  <button>
     +1 
  </button>
  <button>
     -1 
  </button>
  <button
    aria-label="btn-reset"
  >
    Reset
  </button>
</div>
`;
